// @azure/arm-datafactory/esm/operations/activityRuns.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as tslib_1 from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/activityRunsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ActivityRuns. */
var ActivityRuns = /** @class */ (function () {
    /**
     * Create a ActivityRuns.
     * @param {DataFactoryManagementClientContext} client Reference to the service client.
     */
    function ActivityRuns(client) {
        this.client = client;
    }
    ActivityRuns.prototype.queryByPipelineRun = function (resourceGroupName, factoryName, runId, filterParameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            factoryName: factoryName,
            runId: runId,
            filterParameters: filterParameters,
            options: options
        }, queryByPipelineRunOperationSpec, callback);
    };
    return ActivityRuns;
}());
export { ActivityRuns };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var queryByPipelineRunOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/pipelineruns/{runId}/queryActivityruns",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.factoryName,
        Parameters.runId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "filterParameters",
        mapper: tslib_1.__assign({}, Mappers.RunFilterParameters, { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ActivityRunsQueryResponse
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
