// @azure/arm-datafactory/esm/operations/pipelineRuns.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as tslib_1 from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/pipelineRunsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a PipelineRuns. */
var PipelineRuns = /** @class */ (function () {
    /**
     * Create a PipelineRuns.
     * @param {DataFactoryManagementClientContext} client Reference to the service client.
     */
    function PipelineRuns(client) {
        this.client = client;
    }
    PipelineRuns.prototype.queryByFactory = function (resourceGroupName, factoryName, filterParameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            factoryName: factoryName,
            filterParameters: filterParameters,
            options: options
        }, queryByFactoryOperationSpec, callback);
    };
    PipelineRuns.prototype.get = function (resourceGroupName, factoryName, runId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            factoryName: factoryName,
            runId: runId,
            options: options
        }, getOperationSpec, callback);
    };
    PipelineRuns.prototype.cancel = function (resourceGroupName, factoryName, runId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            factoryName: factoryName,
            runId: runId,
            options: options
        }, cancelOperationSpec, callback);
    };
    return PipelineRuns;
}());
export { PipelineRuns };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var queryByFactoryOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/queryPipelineRuns",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.factoryName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "filterParameters",
        mapper: tslib_1.__assign({}, Mappers.RunFilterParameters, { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.PipelineRunsQueryResponse
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/pipelineruns/{runId}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.factoryName,
        Parameters.runId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.PipelineRun
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var cancelOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/pipelineruns/{runId}/cancel",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.factoryName,
        Parameters.runId
    ],
    queryParameters: [
        Parameters.isRecursive,
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
