// @azure/arm-datafactory/esm/operations/triggerRuns.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as tslib_1 from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/triggerRunsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a TriggerRuns. */
var TriggerRuns = /** @class */ (function () {
    /**
     * Create a TriggerRuns.
     * @param {DataFactoryManagementClientContext} client Reference to the service client.
     */
    function TriggerRuns(client) {
        this.client = client;
    }
    TriggerRuns.prototype.queryByFactory = function (resourceGroupName, factoryName, filterParameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            factoryName: factoryName,
            filterParameters: filterParameters,
            options: options
        }, queryByFactoryOperationSpec, callback);
    };
    return TriggerRuns;
}());
export { TriggerRuns };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var queryByFactoryOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/queryTriggerRuns",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.factoryName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "filterParameters",
        mapper: tslib_1.__assign({}, Mappers.RunFilterParameters, { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.TriggerRunsQueryResponse
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
